geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyAggregatingIterator.scala

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2013-2015 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0 which
+ * accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ */
+
+package org.locationtech.geomesa.accumulo.iterators
+
+import java.util.{Collection => jCollection, Map => jMap}
+
+import com.typesafe.scalalogging.slf4j.Logging
+import org.apache.accumulo.core.client.IteratorSetting
+import org.apache.accumulo.core.data.{Range => aRange, _}
+import org.apache.accumulo.core.iterators.{IteratorEnvironment, SortedKeyValueIterator}
+import org.geotools.filter.text.ecql.ECQL
+import org.locationtech.geomesa.features.kryo.{KryoBufferSimpleFeature, KryoFeatureSerializer}
+import org.locationtech.geomesa.filter.factory.FastFilterFactory
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.Filter
+
+import scala.collection.JavaConverters._
+
+/**
+ * Aggregating iterator - only works on kryo-encoded features
+ */
+abstract class KryoLazyAggregatingIterator[T <: AnyRef { def isEmpty: Boolean; def clear(): Unit }]
+    extends SortedKeyValueIterator[Key, Value] {
+
+  import KryoLazyAggregatingIterator._
+
+  var sft: SimpleFeatureType = null
+  var source: SortedKeyValueIterator[Key, Value] = null
+
+  private var validate: (SimpleFeature) => Boolean = null
+
+  // our accumulated result
+  private var result: T = _
+
+  protected var topKey: Key = null
+  private var topValue: Value = new Value()
+  private var currentRange: aRange = null
+
+  private var reusableSf: KryoBufferSimpleFeature = null
+
+  // server-side deduplication - not 100% effective, but we can't dedupe client side as we don't send ids
+  private val idsSeen = scala.collection.mutable.HashSet.empty[String]
+  private var maxIdsToTrack = -1
+
+  override def init(src: SortedKeyValueIterator[Key, Value],
```

Comment:
@Afterglow375 I changed the signature to make this work better. Take a look (in fcr_z3_polys) and let me know.